packages/siwe/src/provider.tsx

```diff
@@ -64,28 +146,67 @@ export const SiweProvider = ({ children, backendUrl, api }: SiweProviderProps) =
         chainId,
         nonce,
       })
-      const signature = await signer.signMessage(message.prepareMessage())
 
-      await signIn({ signature, message })
+      setLoading(true)
+      const signature = await signer.signMessage(message.prepareMessage()).catch((err) => {
+        setError(err)
+        setLoading(false)
+      })
+      if (!signature) {
+        return
+      }
 
-      void getAuth().then((res) => (res.loggedIn ? setLoggedIn(true) : undefined))
+      try {
+        await signIn({ signature, message })
+      } catch (err) {
+        return setError(err)
+      }
+
+      if (signature === '0x') {
+        localStorage.setItem('siweMessage', JSON.stringify(message))
+        return createMultiSigListener({
+          message,
+        })
+      }
+
+      void getAuthRequestHandler()
     },
-    [account, chainId, library, getAuth, getNonce, signIn]
+    [account, chainId, library]
   )
 
   const handleSignOut = useCallback(async () => {
     if (!account || !chainId) {
       return
     }
+    await signOutRequestHandler()
+  }, [account, chainId])
+
+  const createMultiSigListener = async ({ message }: { message: SiweMessage }) => {
+    const gnosisSafeContract = new Contract(message.address, new utils.Interface(GNOSIS_SAFE_ABI), library)
+
+    let getMessageHash = localStorage.getItem('getMessageHash')
+    if (!getMessageHash) {
+      getMessageHash = await gnosisSafeContract.getMessageHash(utils.hashMessage(message.prepareMessage()))
+      localStorage.setItem('getMessageHash', getMessageHash as string)
+    }
 
-    await signOut()
-    setLoggedIn(false)
-  }, [setLoggedIn, account, chainId, signOut])
+    const onMultiSigSigned = async () => {
+      gnosisSafeContract.removeListener(gnosisSafeContract.filters.SignMsg(getMessageHash), onMultiSigSigned)
```

Comment:
Check if `gnosisSafeContract.filters.SignMsg(getMessageHash)` should not be extracted here. Also doesn't `once` remove listener automatically?